stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -1197,6 +1197,150 @@ public func _TFCCreateTensorHandleFromC(
+                              _ attrName: UnsafePointer<Int8>,
+                              _ value: Array<Double>) {
+  value.map(Float.init).withUnsafeBufferPointer { buffer in
+    TFE_OpSetAttrFloatList(op, attrName, buffer.baseAddress, Int32(buffer.count))
+  }
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrFloatArray")
+func _TFCOpSetAttrFloatArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<Float>) {
+  value.withUnsafeBufferPointer { buffer in
+    TFE_OpSetAttrFloatList(op, attrName, buffer.baseAddress, Int32(buffer.count))
+  }
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrTypeArray")
+func _TFCOpSetAttrTypeArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<TensorDataType>) {
+  value.withUnsafeBufferPointer { buffer in
+    buffer.withMemoryRebound(to: TF_DataType.self) { reboundBuffer in
+      TFE_OpSetAttrTypeList(op, attrName, reboundBuffer.baseAddress,
+                            Int32(reboundBuffer.count))
+    }
+  }
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrTensorShapeArray")
+func _TFCOpSetAttrTensorShapeArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<TensorShape>,
+                             _ status: CTFStatus) {
+  let flattenedDims = value.flatMap { $0.dimensions.map(Int64.init) }
+  let ranks = value.map { $0.rank }
+  setAttrShapeList(op: op, attrName: attrName, flattenedDims: flattenedDims,
+                   ranks: ranks, status: status)
+}
+
+@usableFromInline
+@_silgen_name("_swift_tfc_OpSetAttrOptionalTensorShapeArray")
+func _TFCOpSetAttrOptionalTensorShapeArray(_ op: OpaquePointer,
+                             _ attrName: UnsafePointer<Int8>,
+                             _ value: Array<TensorShape?>,
+                             _ status: CTFStatus) {
+  let flattenedDims = value.flatMap { (tensorShapeOpt) -> [Int64] in
+    if let tensorShape = tensorShapeOpt {
+      return tensorShape.dimensions.map(Int64.init)
+    }
+    return []
+  }
+  let ranks = value.map { tensorShapeOpt -> Int32 in
+    if let tensorShape = tensorShapeOpt {
+      return tensorShape.rank
+    }
+    return -1
+  }
+  setAttrShapeList(op: op, attrName: attrName, flattenedDims: flattenedDims,
+                   ranks: ranks, status: status)
+}
+
+/// Given dimensions and ranks in the form described below, makes the
+/// appropriate call to TFE_OpSetAttrShapeList(op, attrName, ..., status).
+///
+/// - Parameters
+///   - flattenedDims: all the shapes' dimensions concatenated together in
+///     order
+///   - ranks: all the shapes' ranks (-1 denotes unknown rank)
+func setAttrShapeList(op: OpaquePointer, attrName: UnsafePointer<Int8>,
+                      flattenedDims: Array<Int64>, ranks: Array<Int32>,
```

Comment:
Can we replace `op: OpaquePointer` with `op: CTFEOp`?

It's the same from compiler perspective, but serves as a better documentation for human.

Reply:
done